Add number & range Usercss variables (#492)

Mottie opened the pull request:
Adds new usercss variables & behavior:

* Add number & range variables:
  * `@var number some-number "Number" [10, 0, 20, 1, "px"]` (`[default, min, max, step, units]`).
  * `@var range  some-range  "Range (0-10)"  [5, 0, 10, 1, "em"]` (`[default, min, max, step, units]`)
  * The `min`, `max`, `step` and `units` are optional; but added sequentially.
* Focusing on a text input will select all content.
* Default select option is set by including a `*` at the end of the key, e.g.
  * `@var select aSelect "Font"   ["Arial", "Consolas*", "Times New Roman"]`
  * ```
    @var select aSelect "Font" {
      "Arial": "arial",
      "Consolas*": "consolas",
      "Times": "Times New Roman"
    }
    ```

------

Example UserCSS:

```css
/* ==UserStyle==
@namespace      stylus#492
@name           new variables
@version        1.0.0
@description    Testing new variables
@author         Me
@var text   aTxt    "Text"    "Blah"
@var number aNum    "Margin"  [1, 0, 10, 0.5, "em"]
@var select aSelect "Font"    ["Arial", "Consolas*", "Times New Roman"]
@var range  aRange0 "Opacity" [1, 0, 1, 0.1]
@var range  aRange1 "Width"   [5, 1, 10, 1, "px"]
@var range  aRange2 "Height"  [11, 10, "vh"]
@preprocessor uso
==/UserStyle== */
@-moz-document domain("google.com") {
  body > div:last-child {
    content: '/*[[aTxt]]*/';
    margin: /*[[aNum]]*/;
    font-family: '/*[[aSelect]]*/';
    opacity: /*[[aRange0]]*/;
    width: /*[[aRange1]]*/;
    height: /*[[aRange2]]*/;
  }
}
```

![](https://user-images.githubusercontent.com/136959/45601540-35b41600-b9d4-11e8-8229-939460c3b5e7.png)

Mottie commented:
I didn't include the scrollwheel enhancement that was requested.

tophf commented on js/usercss.js:
```diff
@@ -177,6 +177,7 @@ var usercss = (() => {
     result.label = state.value;
 
     const {re, type, text} = state;
+    let deflt;
```
srsly?
pls rnme it

tophf commented on manage/config-dialog.js:
```diff
@@ -297,6 +314,10 @@ function configDialog(style) {
     }
   }
 
+  function focusInput(event) {
```
This is a misleading function name: it doesn't focus the input. Maybe onInputFocused or selectAllOnFocus to match the existing naming scheme

tophf commented on js/usercss.js:
```diff
@@ -195,11 +196,40 @@ var usercss = (() => {
-          result.options = state.value.map(text => createOption(text));
+          result.options = state.value.map(text => {
+            if (text.endsWith('*')) {
+              text = text.replace(/\*$/, '');
```
`text = text.slice(0, -1).trim()`

tophf commented on js/usercss.js:
```diff
@@ -195,11 +196,40 @@ var usercss = (() => {
+          result.options = Object.keys(state.value).map(k => {
+            if (k.endsWith('*')) {
+              state.value[k] = state.value[k].replace(/\*$/, '');
+              k = k.replace(/\*$/, '');
```
use `.slice(0, -1).trim()` in both cases

tophf replied:
er, I guess trim() isn't needed because some values may really need those trailing spaces

Mottie commented:
Sorry, long day yesterday... 

So, what are everyone's thoughts on adding a range variable? Might it even be necessary (vs number type with min/max/step)? It is good now on how it's set in an array? Does it need a min and max value positioned around the slider?

eight04 commented:
> It is good now on how it's set in an array?

I prefer to set the default value as the first item.

Should we make `number` accept an array as well? Some properties might be set to `null` e.g. `[5, 0, null, null]`.

eight04 commented:
> Does it need a min and max value positioned around the slider?

I think we need some screenshots.

Mottie commented:
So `[default, min, max, step]`? I'll take another look later when I get time, unless you want to make the changes to this branch yourself? 😸

Mottie commented:
Ok, the range settings have been modified and now the `min`, `max` and `step` are optional; but still set sequentially. I've updated the usercss in the original post with a screenshot.

I added an outline to the current range value, and the range sliders don't exactly line up when no `max` is set... oh well.

Mottie commented:
I also discovered that adding decimal points without a leading zero was causing JSON parsing errors

```json
[.5, 0, 1, .1] // invalid JSON
```

```json
[0.5, 0, 1, 0.1] // valid!
```

Mottie commented:
DOH, and I forgot to make the number variable to also accept a min, max & step... I'll get to it tomorrow.

Mottie commented:
Updated UserCSS example & screenshot in original post.

Mottie commented:
Hmm, also what do you think about including a label with the default values, e.g.

```
@var range some-range "Range (0-10)" ['5px', 0, 10, 1]
```

So when the value is set, the `px` is included in the setting.

That would save the hassle of using `/*[[some-range]]*/px` which *shows as an error in CSSLint*, or `calc(/*[[some-range]]*/ * 1px)`.

tophf commented:
Indeed, we can apply `uso` preprocessor using the default or current variable values before sending the text to csslint (or try allowing uso var comments everywhere which is IMO meh...). Still, it might make sense to have the unit specified (optionally) in the variable declaration separately so it's automatically added to the number when used.

eight04 commented:
> have the unit specified (optionally) in the variable declaration separately

Something like `[5, 0, 10, 1, 'px']`?

Mottie commented:
Changes:

* Removed range values from number type inputs.
* Range & number inputs will now accept a "units" value: `[1, 0, 10, 1, "em"]`.
* Range & number values are now properly validated; any non-numeric values will be removed from the range.
* Range current value will also include units.
* Manually inputted "number" values will be clamped to the range & adjusted to the step value (if any).

Mottie commented:
> We could allow the unit to be anywhere

I made it so the unit label could be added anywhere *after* the default value. And if multiple labels/text array items were included, only the first is assigned to the label and the rest are stripped out.

```
[1, "px", 0, 10, "foo", 2] => [0, 10, 2] ([min, max, step] with default = 1 & units = "px")
```

stonecrusher commented:
Can the end-user edit the unit, so he can decide if pixel or percent?

Mottie commented:
> Can the end-user edit the unit, so he can decide if pixel or percent?

No, only the author... or if the user decides to edit it.

Mottie commented:
I believe this is ready to be merged. @eight04 can you review it to see if there are any issues. Will you be able to adapt these updates to the UserCSS parser as well?

eight04 commented:
> @eight04 can you review it to see if there are any issues

Sure. I'll look into this tomorrow.

> Will you be able to adapt these updates to the UserCSS parser as well?

You mean `usercss-meta`? I think it should follow the usercss spec. We can release a new version after the spec (the wiki) is updated. You can file a PR now or I may find some time to file a PR after the spec is updated.

And we may start working on #496 after that.

Mottie commented:
I've already completed the wiki page, I was going to update it after the PR was merged, but I can do it now.

Mottie commented:
Ok done... I renamed the page to [UserCSS authors](https://github.com/openstyles/stylus/wiki/UserCSS-authors), and now the [UserCSS](https://github.com/openstyles/stylus/wiki/UserCSS) page contains basic infomation for non-developers.

eight04 commented on manage/config-dialog.js:
```diff
@@ -286,7 +314,26 @@ function configDialog(style) {
+    } else if (value > max) {
+      return max;
+    }
+    const remainder = value % (step || 1);
```
This won't work with floats because of the precision. If we really need this, we have to upscale the value to an integer then downscale them after the calculation.

Mottie replied:
I tested this using a step of 0.5...

```
1.9 % 0.5
// => 0.3999999999999999
// but
1.9 - (1.9 % 0.5)
// => 1.5

So the precision should cancel out...

eight04 replied:
0.5 can be saved as float without losing precision. Try `1 - 1 % 0.1`, because 0.1 float number is actually slightly larger than actual 0.1, the result of 1 % 0.1 is slightly smaller than 0.1, then 1 - (1 % 0.1) becomes 0.9.

Mottie replied:
Oh yeah, good example.... so maybe just leave it out then?

tophf replied:
Detect the precision of `step` and round the result to it?
Something like stepPrecision = stepStr.match(/\.(\d+?[1-9])0*\b|$/)[1].length;
N.B. A reliable maximum is 15 digits for IEEE754 floats used in JS including all the digits, not just the fractional part.

eight04 replied:
We don't have to round the value for users IMHO. If the author made some validations and the user don't care about the error, we can accept it as well.

The main problem is how to tell the user that the input is invalid. I think we can:

1. Add some styles to `input:invalid`.
2. Call `input.reportValidity()` on `change` event to show the error message.

Mottie replied:
I changed the method of adjusting the value to the nearest step below the value... not sure what to call it, clamp to step? Anyway, it should work now; it might still need the `.toFixed()` function for cleanup...

eight04 commented:
This is the current UI:
![image](https://user-images.githubusercontent.com/1324510/45592080-7e4cde80-b996-11e8-98a2-047030b26661.png)

I think it would look better if we replace those min/max value and the current value text box with a simple range tooltip. Something similar to this:
https://codepen.io/eight04/pen/dqqpYy

Mottie commented on js/usercss.js:
```diff
@@ -195,11 +196,42 @@ var usercss = (() => {
         if (Array.isArray(state.value)) {
-          result.options = state.value.map(text => createOption(text));
+          result.options = state.value.map(text => {
+            if (text.endsWith('*')) {
```
> Move them to a function:

This method removes the setting of the default value, we're not setting `result.default` anymore. So either we remove the `defaultValue` or skip moving the extraction method.

> It is not clear that whether we should write

I wrote that it is added to the end of the *key* name... so the second example is correct. I think I need to include that example in the docs.

Mottie commented on js/usercss.js:
```diff
@@ -195,11 +196,42 @@ var usercss = (() => {
-          result.options = Object.keys(state.value).map(k => createOption(k, state.value[k]));
+          result.options = Object.keys(state.value).map(k => {
+            if (k.endsWith('*')) {
+              state.value[k] = state.value[k].slice(0, -1);
```
`slice(0, -1)` removes the `*` from the end of the string. It's not an error.

eight04 replied:
If someone writes:
```
@var select my-select "my select" {
  "foo": "A foo",
  "bar*": "A bar"
}
```
This line would convert `"A bar"` into `"A ba"`.

Mottie replied:
DOH, you're right.

Mottie commented on js/usercss.js:
```diff
@@ -195,11 +196,42 @@ var usercss = (() => {
+            return createOption(k, state.value[k]);
+          });
+        }
+        result.default = typeof defaultValue !== 'undefined' ? defaultValue : (result.options[0] || {}).name || '';
```
I was checking type in case the default value was set to zero.

eight04 replied:
The default value of `@var select` can't be zero.

eight04 replied:
This is not the correct way to set the default value of `@var select`. The default value should be `option.name` that `option` is taken from `result.options`. We should use a `defaultIndex`:
```diff
diff --git a/js/usercss.js b/js/usercss.js
index aef8314..fff41a0 100644
--- a/js/usercss.js
+++ b/js/usercss.js
@@ -195,6 +195,7 @@ var usercss = (() => {
         state.errorPrefix = 'Invalid JSON: ';
         parseJSONValue(state);
         state.errorPrefix = '';
+        let defaultIndex = 0;
         if (Array.isArray(state.value)) {
           result.options = state.value.map(text => {
             if (text.endsWith('*')) {
@@ -204,16 +205,16 @@ var usercss = (() => {
             return createOption(text);
           });
         } else {
-          result.options = Object.keys(state.value).map(k => {
+          result.options = Object.keys(state.value).map((k, i) => {
             if (k.endsWith('*')) {
               state.value[k] = state.value[k].slice(0, -1);
               k = k.slice(0, -1);
-              defaultValue = k;
+              defaultIndex = i;
             }
             return createOption(k, state.value[k]);
           });
         }
-        result.default = typeof defaultValue !== 'undefined' ? defaultValue : (result.options[0] || {}).name || '';
+        result.default = (result.options[defaultIndex] || {}).name || '';
         break;
       }
 

```

Mottie replied:
I rewrote this section... I might have missed this change.

Mottie commented on js/usercss.js:
```diff
@@ -195,11 +196,42 @@ var usercss = (() => {
+      case 'number':
+      case 'range': {
+        state.errorPrefix = 'Invalid JSON: ';
+        parseJSONValue(state);
```
I wasn't sure how to do that... we'd need to differentiate when the number includes a min, max, step and units versus only a default value. Got any advice on that?

eight04 replied:
```diff
diff --git a/js/usercss.js b/js/usercss.js
index aef8314..4b079fb 100644
--- a/js/usercss.js
+++ b/js/usercss.js
@@ -218,6 +218,15 @@ var usercss = (() => {
       }
 
       case 'number':
+        if (state.text[state.re.lastIndex] !== '[') {
+          parseNumber(state);
+          result.default = state.value;
+          result.range = [];
+          result.units = '';
+          break;
+        }
+        // pass through
+
       case 'range': {
         state.errorPrefix = 'Invalid JSON: ';
         parseJSONValue(state);

```
Actually, I think we can delay this to `usercss-meta`. Let's fix other bugs first.

Mottie replied:
I didn't make this change... it's still set to use an array.

Mottie commented on js/usercss.js:
```diff
@@ -195,11 +196,42 @@ var usercss = (() => {
+        state.errorPrefix = '';
+        // [default, start, end, step, units] (start, end, step & units are optional)
+        if (Array.isArray(state.value) && state.value.length) {
+          result.default = parseFloat(state.value.shift());
```
You're right, I forgot I left that in because of my original idea of including the units along with the value.

Mottie commented on js/usercss.js:
```diff
@@ -195,11 +196,42 @@ var usercss = (() => {
+          result.default = parseFloat(state.value.shift());
+          const nonDigit = /[^\d.+-]/;
+          // label may be placed anywhere after default value
+          const labelIndex = state.value.findIndex(item => nonDigit.test(item));
```
Nice! I'll keep the replace digits function for the units though, in case someone added a number.

Mottie commented on js/usercss.js:
```diff
@@ -195,11 +196,42 @@ var usercss = (() => {
+          const labelIndex = state.value.findIndex(item => nonDigit.test(item));
+          // but should not contain any numbers '4px' => 'px'
+          result.units = labelIndex < 0 ? '' : state.value.splice(labelIndex, 1)[0].toString().replace(/[\d.+-]/g, '');
+          result.range = state.value.filter(item => !nonDigit.test(item));
```
You mean have the `@var` set as an object instead of an array? The array is a lot less typing.

```
@var range    imgHeight  "Max image height" [50, 10, 200, 10, "px"]
```

versus

```
@var range    imgHeight  "Max image height" {
  "default": 50,
  "min": 10,
  "max": 200,
  "step": 10,
  "units": "px"
}
```

Mottie replied:
Nice... I cleaned it all up using this method.

Mottie commented on manage/config-dialog.js:
```diff
@@ -201,6 +201,27 @@ function configDialog(style) {
+    if (typeof va.range[1] !== 'undefined') {
+      options.max = dataset.max = va.range[1];
+    }
+    if (va.range[2]) {
```
The above suggestions are good, but this one I don't think we want the user to set a step of zero.

eight04 replied:
This should be OK then.

Mottie commented:
Yes that would look nice, but the user would have to click on the slider to see the current value. I am indifferent to including the min/max values, but I think we *must* have the current value visible; also, accessibility.

eight04 commented:
IMHO, `range` was not designed to input a precise value.

I would vote not showing the min/max values so those sliders are aligned. And we may want to alter the style of the `.current-value` box. It looks like a text input now.

tophf commented:
It's quite plausible to assume a range can be used to input precise values 50% of the time. That is without the actual usage statistics any number is good, 50 being the average. We need to round the value to avoid the inherent problem of all engines that use IEEE754 floats so that we don't end up showing 6.9999999999997 etc.

Mottie commented:
I kinda mashed all the fixes into one big commit - sorry about that. 

* Added `extractDefault` function.
* All number & range values must now be a number, or they are ignored; the first string value in the array becomes the units value.
* Removed the min, max & outline from the range current value.
* Used a different method to round to the nearest step value:

  ```js
  const inv = 1 / (step || 1);
  return typeof step !== 'undefined' ? Math.floor(inv * value) / inv : value;
  ```

Mottie commented on js/usercss.js:
```diff
@@ -177,6 +177,7 @@ var usercss = (() => {
     result.label = state.value;
 
     const {re, type, text} = state;
+    let defaultValue;
```
Code changed, no longer applicable.

Mottie commented on manage/config-dialog.js:
```diff
@@ -258,16 +279,23 @@ function configDialog(style) {
+            oninput: updateVarOnInput,
+            onfocus: selectAllOnFocus,
+          };
+          if (va.type === 'range' || va.type === 'number') {
```
Code changed, no longer applicable.

Mottie commented on js/usercss.js:
```diff
@@ -578,6 +613,11 @@ var usercss = (() => {
       va[value] = colorConverter.format(colorConverter.parse(va[value]), 'rgb');
+    } else if (
+      (va.type === 'number' || va.type === 'range') &&
+      !(typeof va[value] === 'number' || Array.isArray(va.range))
```
`va.range` is no longer defined.

Mottie commented on manage/config-dialog.js:
```diff
@@ -201,6 +201,27 @@ function configDialog(style) {
+  function handleRangeAndNumberInputs(va, options) {
+    const dataset = {};
+    options.value = va.default;
+    if (typeof va.range[0] !== 'undefined') {
```
Code changed, no longer applicable.

eight04 commented on manage/config-dialog.css:
```diff
@@ -98,6 +98,25 @@
+  line-height: 22px;
+}
+
+.current-range:before {
```
We don't need this anymore?

eight04 commented on manage/config-dialog.css:
```diff
@@ -98,6 +98,25 @@
+  padding-right: .25em;
+}
+
+.current-range:after {
```
Same as above.

eight04 commented on manage/config-dialog.js:
```diff
@@ -258,16 +258,35 @@ function configDialog(style) {
           break;
+
+        default: {
+          const options = {
```
This object could be inlined.

Mottie commented on js/usercss.js:
```diff
@@ -194,12 +194,52 @@ var usercss = (() => {
+        if (Array.isArray(state.value) && state.value.length) {
+          // label may be placed anywhere
+          result.units = (state.value.find(i => typeof i === 'string') || '').replace(/[\d.+-]/g, '');
+          const range = state.value.filter(i => typeof i === 'number');
```
Nevermind, I just updated the wiki.

narcolepticinsomniac replied:
First step is admitting you have a problem. =)

Mottie replied:
It's hard to code while holding a baby LOL... but now she's asleep, so I can type again! 😛

tophf commented on manage/config-dialog.js:
```diff
@@ -197,8 +197,12 @@ function configDialog(style) {
   }
 
+  function isNullOrUndefined(value) {
+    return value === null || value === undefined;
```
That would require adding an eslint exception. Anyway we decided to never rely on the implicit quirks on loose comparison so it'd really weird to use it. 

I also think this isNullOrUndefined looks like a kludge which could be implemented better e.g. `!isNaN(Number(va.min))` - this can be extracted to a function like isNumber, of course. For `step` and default value `isFinite` should be added to the check, otherwise `1e11111` will produce `Infinity` which is only valid for max (and `-Infinity` is valid for `min`).

Mottie replied:
I ended up with:

```js
function isNumber(value) {
  return value !== '' && !isNaN(Number(value));
}
```

since `Number('')` returns `0`.

eight04 replied:
`Number(null)` also returns 0. Use something like `typeof value === "number" && !isNaN(value)`

Mottie replied:
But `typeof '1' === 'number'` is `false`. So maybe just include a `null` check.

Mottie commented on manage/config-dialog.js:
```diff
@@ -317,17 +328,23 @@ function configDialog(style) {
-    const min = va.min || 0;
-    const max = va.max || 100;
+    const min = typeof va.min === 'undefined' ? 0 : va.min;
+    const max = typeof va.max === 'undefined' ? 100 : va.max;
```
Ok for min, but an undefined max is set to `100`.

eight04 replied:
Why? I don't think authors would expect that there is a default constraint of the max value.

Mottie replied:
Ack, you're right... I thought I noticed it was set that way by the browser. Fixed now.

Mottie commented on manage/config-dialog.js:
```diff
@@ -317,17 +328,23 @@ function configDialog(style) {
+    const precision = (step.toString().split('.')[1] || 0).length + 1;
+    const inv = 1 / step;
+    // ECMA-262 only requires a precision of up to 21 significant digits
+    return Number((Math.floor(inv * value) / inv).toPrecision(precision > 21 ? 21 : precision));
```
Good work on the method; I was almost there 😉.

I ended up running the returned value through the clamp function, but making sure we don't get an infinite loop.

eight04 commented on js/usercss.js:
```diff
@@ -235,7 +263,7 @@ var usercss = (() => {
 
       default: {
-        // text, color
+        // text, color, number
```
This change is misleading.

eight04 commented on manage/config-dialog.js:
```diff
@@ -197,6 +197,10 @@ function configDialog(style) {
     renderValues();
   }
 
+  function isNumber(value) {
```
Since this would detect numeric string, maybe it should be renamed to `isNumeric`.

Mottie replied:
I didn't make this change, but instead changed the function - see comment below.

eight04 commented on manage/config-dialog.js:
```diff
@@ -197,6 +197,10 @@ function configDialog(style) {
   }
 
+  function isNumber(value) {
+    return value !== null && value !== '' && !isNaN(Number(value));
```
This would fail with:
```js
const foo = {
  toString: () => ""
};
isNumber(foo); // true
```

tophf replied:
And how would you provide a toString function in usercss meta header?

eight04 replied:
You can't, but it doesn't mean the function is doing the right thing.

tophf replied:
Do you mean it can fail on the real input? Or do you mean it's theoretically wrong? Because I don't really care about the latter.

eight04 replied:
> Do you mean it can fail on the real input?

I'm not sure what is "real input". Currently, this function is only used to check `va.min`/`va.max`/`va.step`.

> do you mean it's theoretically wrong? Because I don't really care about the latter.

Then let's change it to `return typeof value === 'number'`. It is shorter and will work with `va.min`/`va.max`/`va.step`. Also, it would match the function name so you don't have to rename it to `isNumeric`.

Mottie replied:
I think what @eight04 is trying to say is that `va.min`, `va.max` and `va.step` will always be numeric values since only the units label is the only allowed string value while setting the variable. All other strings are filtered out; therefore, we don't need to use `Number()`.

This means we can change the function to:

```js
function isNumber(value) {
  return typeof value === 'number';
}
```

I've been parsing the values over-cautiously. So I'll switch it to the above function.

tophf replied:
There's no need for a function then...

Mottie replied:
Ok, all number checks are inline now.

eight04 commented on manage/config-dialog.js:
```diff
@@ -286,7 +316,33 @@ function configDialog(style) {
   function updateVarOnChange() {
-    this.va.value = this.type !== 'checkbox' ? this.value : this.checked ? '1' : '0';
+    if (this.type === 'number') {
+      this.value = this.va.value = clampValue(this.value, this.va);
```
I think resetting the value while the user is typing is confusing. Maybe we should only reset the input value on the `blur` event?
![test](https://user-images.githubusercontent.com/1324510/45734409-08a37580-bc17-11e8-87da-9d143338e736.gif)

tophf replied:
Or show a small tooltip that informs the value was reset.

Mottie replied:
I changed it to only clamp the value on blur, but now the raw value is saved in the live preview... I guess that's okay because it changes once the input is blurred.

Mottie replied:
Actually, if you immediately close the popup, the invalid value remains.

Mottie replied:
Even if we use a timer to reset the value, it'll remain if the user closes the dialog/popup.

tophf replied:
I think this is bad, and worse than having the value reset forcefully while typing. In colorpicker I simply didn't use an invalid value and nothing was reported upstream, the input element was set into an invalid state which was reddened via CSS. Can you do the same here?

Mottie replied:
Yes, this is the best idea... now implemented.

eight04 commented on manage/config-dialog.js:
```diff
@@ -286,7 +316,33 @@ function configDialog(style) {
+  }
+
+  // Clamp input[type=number] to a valid range
+  function clampValue(value, va, break = false) {
```
Syntax error.

Mottie replied:
Yeah, I forgot `break` was a keyword... long day yesterday.

eight04 commented on manage/config-dialog.js:
```diff
@@ -286,7 +316,33 @@ function configDialog(style) {
+    const scale = 10 ** (step.toString().split('.')[1] || '').length;
+    value = Math.round((value * scale) / (step * scale)) * (step * scale) / scale;
+    // clamp modified value; skip step check
+    return clampValue(value, va, true);
```
This is a bad idea. It makes the first min/max check useless. Let's move L341-L343 to L331 so the step validation is run before min/max validation.

Mottie replied:
Yeah, I thought that breaking out of the recursion would be sufficient, but your idea is better.

eight04 commented on manage/config-dialog.js:
```diff
@@ -286,7 +313,34 @@ function configDialog(style) {
+      this.va.value = Number(this.value);
+      $('.current-value', this.closest('.config-range')).textContent = this.va.value + (this.va.units || '');
+    } else {
+      this.va.value = this.type !== 'checkbox' ? this.value : this.checked ? '1' : '0';
```
For `number`, you can update `this.va.value` on change event, but update `input.value` when blurred:
```diff
diff --git a/manage/config-dialog.js b/manage/config-dialog.js
index 34559c4..23431bc 100644
--- a/manage/config-dialog.js
+++ b/manage/config-dialog.js
@@ -320,6 +320,9 @@ function configDialog(style) {
     if (this.type === 'range') {
       this.va.value = Number(this.value);
       $('.current-value', this.closest('.config-range')).textContent = this.va.value + (this.va.units || '');
+    } else if (this.type === "number") {
+      // assign a valid number to va.value
+      this.va.value = clampValue(Number(this.value), this.va);
     } else {
       this.va.value = this.type !== 'checkbox' ? this.value : this.checked ? '1' : '0';
     }
@@ -327,7 +330,8 @@ function configDialog(style) {
 
   // Applied to input[type=number]
   function updateVarOnBlur() {
-    this.value = this.va.value = clampValue(this.value, this.va);
+    // update the value of the input
+    this.value = this.va.value;
   }
 
   // Clamp input[type=number] to a valid range
```

Mottie replied:
Good idea! But, maybe we shouldn't change the `updateVarOnBlur()` to ensure the values are the same.

Mottie replied:
No longer necessary... the value is no longer saved if invalid.

eight04 commented on manage/config-dialog.js:
```diff
@@ -286,7 +313,38 @@ function configDialog(style) {
+      $('.current-value', this.closest('.config-range')).textContent = this.va.value + (this.va.units || '');
+    } else if (this.type === 'number') {
+      // clamp stored value
+      this.va.value = clampValue(this.value, this.va);
```
It should be `Number(this.value)`. `input.value` is always a string.

Mottie replied:
DOH, fixed.

eight04 commented on manage/config-dialog.js:
```diff
@@ -286,7 +313,38 @@ function configDialog(style) {
+  // Applied to input[type=number]
+  function updateVarOnBlur() {
+    // clamp value visible to user
+    this.value = this.va.value = clampValue(this.value, this.va);
```
Same as above.

Mottie replied:
Function removed.

eight04 commented on manage/config-dialog.js:
```diff
@@ -307,6 +369,12 @@ function configDialog(style) {
       } else if (va.type === 'checkbox') {
         va.input.checked = Number(value);
+      } else if (va.type === 'range') {
+        const span = $('.current-value', va.input.closest('.config-range'));
```
There are various places that would update `.current-value` box for the range input. Move them to a function. And we can also update it on input event so the value is updated immediately:
```diff
diff --git a/manage/config-dialog.js b/manage/config-dialog.js
index 697e174..0615cb1 100644
--- a/manage/config-dialog.js
+++ b/manage/config-dialog.js
@@ -315,7 +315,7 @@ function configDialog(style) {
   function updateVarOnChange() {
     if (this.type === 'range') {
       this.va.value = Number(this.value);
-      $('.current-value', this.closest('.config-range')).textContent = this.va.value + (this.va.units || '');
+      updateRangeCurrentValue(this.va, this.va.value);
     } else if (this.type === 'number') {
       // clamp stored value
       this.va.value = clampValue(this.value, this.va);
@@ -347,7 +347,15 @@ function configDialog(style) {
     return value;
   }
 
+  function updateRangeCurrentValue(va, value) {
+    const span = $('.current-value', va.input.closest('.config-range'));
+    span.textContent = value + (va.units || '');
+  }
+
   function updateVarOnInput(event, debounced = false) {
+    if (event.target.type === 'range') {
+      updateRangeCurrentValue(event.target.va, event.target.value);
+    }
     if (debounced) {
       event.target.dispatchEvent(new Event('change', {bubbles: true}));
     } else {
@@ -370,11 +378,8 @@ function configDialog(style) {
       } else if (va.type === 'checkbox') {
         va.input.checked = Number(value);
       } else if (va.type === 'range') {
-        const span = $('.current-value', va.input.closest('.config-range'));
         va.input.value = value;
-        if (span) {
-          span.textContent = value + (va.units || '');
-        }
+        updateRangeCurrentValue(va, value);
       } else {
         va.input.value = value;
       }
```
![test](https://user-images.githubusercontent.com/1324510/45755323-af583800-bc50-11e8-8ac5-af824a845221.gif)

Mottie replied:
Done.

eight04 commented on manage/config-dialog.js:
```diff
@@ -286,7 +312,41 @@ function configDialog(style) {
+      updateRangeCurrentValue(this.va, this.va.value);
+    } else if (this.type === 'number') {
+      const value = Number(this.value);
+      if (clampValue(value, this.va) === value) {
```
Good idea, but it should be `String(clampValue(value, this.va)) === this.value`. Or it would match invalid values for some edge cases e.g. the input is empty `''`, has leading zeros `01`, or has meaningless decimal `1.0`.

tophf replied:
Stringified comparison won't work if the style specifies `1e2` instead of `100`. It's also a valid JSON number.

eight04 replied:
You're right. Users can also input `1e2` directly. Instead of changing it to `String(clampValue(value, this.va)) === this.value`, I think we shouldn't re-render the input while the input is focused:
```diff
diff --git a/manage/config-dialog.js b/manage/config-dialog.js
index efd720b..65734a5 100644
--- a/manage/config-dialog.js
+++ b/manage/config-dialog.js
@@ -363,6 +363,9 @@ function configDialog(style) {
 
   function renderValues(varsToRender = vars) {
     for (const va of varsToRender) {
+      if (va.input === document.activeElement) {
+        continue;
+      }
       const value = isDefault(va) ? va.default : va.value;
       if (va.type === 'color') {
         va.input.style.backgroundColor = value;
```

Mottie replied:
Added this last bit of code to ignore the input while active; but it looks a little weird to have an empty input, set to zero in the background, and not show anything. The update won't update until another variable is changed.

eight04 replied:
Without e595f23, invalid values are also left in the input until another variable is changed. Maybe it is a good idea to update the input to match `input.va.value` when blurred?

Mottie replied:
Ok, that works better.

eight04 commented on manage/config-dialog.js:
```diff
@@ -285,8 +312,50 @@ function configDialog(style) {
+      this.va.value = value;
+    }
+    if (e.type === 'blur') {
+      this.value = clamped;
```
* You don't have to re-clamp the value when blurred, because the correct value is already set in the change handler.
* If the input value is invalid, `this.value` will not match `this.va.value` after blurred.
* In fact, we don't even need `clampValue` if we don't want to convert an invalid value to a correct one.

```diff
diff --git a/manage/config-dialog.js b/manage/config-dialog.js
index c314f2e..69cd88a 100644
--- a/manage/config-dialog.js
+++ b/manage/config-dialog.js
@@ -266,7 +266,8 @@ function configDialog(style) {
             onfocus: va.type === 'number' ? selectAllOnFocus : null,
             onblur: va.type === 'number' ? updateVarOnBlur : null,
             onchange: updateVarOnChange,
-            oninput: updateVarOnInput
+            oninput: updateVarOnInput,
+            required: true
           };
           if (typeof va.min === 'number') {
             options.min = va.min;
@@ -312,15 +313,8 @@ function configDialog(style) {
     }
   }
 
-  function updateVarOnBlur(e = {}) {
-    const value = Number(this.value);
-    const clamped = clampValue(value, this.va);
-    if (clamped === value) {
-      this.va.value = value;
-    }
-    if (e.type === 'blur') {
-      this.value = clamped;
-    }
+  function updateVarOnBlur() {
+    this.value = this.va.value;
   }
 
   function updateVarOnChange() {
@@ -328,29 +322,14 @@ function configDialog(style) {
       this.va.value = Number(this.value);
       updateRangeCurrentValue(this.va, this.va.value);
     } else if (this.type === 'number') {
-      updateVarOnBlur.call(this);
+      if (this.reportValidity()) {
+        this.va.value = Number(this.value);
+      }
     } else {
       this.va.value = this.type !== 'checkbox' ? this.value : this.checked ? '1' : '0';
     }
   }
 
-  // Clamp input[type=number] to a valid range
-  function clampValue(value, va) {
-    // Don't restrict to integer values if step is undefined.
-    if (typeof va.step === 'number') {
-      const step = va.step || 1;
-      const scale = 10 ** (step.toString().split('.')[1] || '').length;
-      value = Math.round((value * scale) / (step * scale)) * (step * scale) / scale;
-    }
-    if (typeof va.min === 'number' && value < va.min) {
-      return va.min;
-    }
-    if (typeof va.max === 'number' && value > va.max) {
-      return va.max;
-    }
-    return value;
-  }
-
   function updateRangeCurrentValue(va, value) {
     const span = $('.current-value', va.input.closest('.config-range'));
     if (span) {
```

eight04 replied:
`updateVarOnBlur` is wrong. The correct one:
```js
  function updateVarOnBlur() {
    this.value = isDefault(this.va) ? this.va.default : this.va.value;
  }
```

Mottie replied:
I didn't know about `reportValidity()`... very cool. Of course it doesn't do anything in Firefox.

eight04 replied:
Firefox:
![image](https://user-images.githubusercontent.com/1324510/45809867-e2a3d100-bcfb-11e8-8a3d-8ce482ea2a56.png)
Chrome:
![image](https://user-images.githubusercontent.com/1324510/45809926-01a26300-bcfc-11e8-878e-efde5aee3fb5.png)

eight04 commented:
Can we merge this? Before working on #496, we should merge this or we will get a bunch of merge conflicts.

Mottie commented:
DOH, I should have squashed this PR! 😿

eight04 commented:
> DOH, I should have squashed this PR!

I guess we should always squash PRs and clarify the changes in the commit message. It should help writing the changelog.
